options/logger.go

```diff
@@ -77,95 +77,120 @@ func (f RawLoggerConfigFormat) Validate() error {
+	if len(*lc) > 0 {
+		return *lc, nil
+	}
+
+	return json.Marshal([]byte(*lc))
+}
+
+func (lc *RawLoggerConfig) UnmarshalJSON(b []byte) error {
+	*lc = b
+	return nil
+}
+
+func (lc RawLoggerConfig) MarshalBSON() ([]byte, error) {
+	if len(lc) > 0 {
+		return lc, nil
+	}
+
+	return bson.Marshal([]byte(lc))
+}
+
+func (lc *RawLoggerConfig) UnmarshalBSON(b []byte) error {
+	*lc = b
+	return nil
 }
 
 // LoggerConfig represents the necessary information to construct a new grip
-// send.Sender. LoggerConfig implements the bson.MarshalBSON and
-// json.MarshalJSON interfaces.
+// send.Sender. LoggerConfig implements the json and bson Marshaler and
+// Unmarshaler interfaces.
 type LoggerConfig struct {
-	Type     string                `json:"type" bson:"type"`
-	Format   RawLoggerConfigFormat `json:"format" bson:"format"`
-	Config   []byte                `json:"config" bson:"config"`
-	Registry LoggerRegistry        `json:"-" bson:"-"`
-
+	info     loggerConfigInfo
+	registry LoggerRegistry
 	producer LoggerProducer
 	sender   send.Sender
 }
 
-// Validate ensures the LoggerConfig is valid.
-func (lc *LoggerConfig) Validate() error {
+type loggerConfigInfo struct {
+	Type   string                `json:"type" bson:"type"`
+	Format RawLoggerConfigFormat `json:"format" bson:"format"`
+	Config RawLoggerConfig       `json:"config" bson:"config"`
+}
+
+func (lc *LoggerConfig) validate() error {
 	catcher := grip.NewBasicCatcher()
 
-	catcher.NewWhen(lc.Type == "", "cannot have empty logger type")
-	if len(lc.Config) > 0 {
-		catcher.Add(lc.Format.Validate())
+	catcher.NewWhen(lc.info.Type == "", "cannot have empty logger type")
+	if len(lc.info.Config) > 0 {
+		catcher.Add(lc.info.Format.Validate())
 	}
 
-	if lc.Registry == nil {
-		lc.Registry = globalLoggerRegistry
+	if lc.registry == nil {
+		lc.registry = globalLoggerRegistry
 	}
 
 	return catcher.Resolve()
 }
 
 // Set sets the logger producer and type for the logger config.
 func (lc *LoggerConfig) Set(producer LoggerProducer) error {
-	if lc.Registry == nil {
-		lc.Registry = globalLoggerRegistry
+	if lc.registry == nil {
+		lc.registry = globalLoggerRegistry
 	}
 
-	if !lc.Registry.Check(producer.Type()) {
+	if !lc.registry.Check(producer.Type()) {
 		return errors.New("unregistered logger producer")
 	}
-	lc.Type = producer.Type()
+
+	lc.info.Type = producer.Type()
 	lc.producer = producer
 
 	return nil
 }
 
-// GetProducer returns the underlying logger producer for this logger config,
+// Producer returns the underlying logger producer for this logger config,
 // which may be nil.
-func (lc *LoggerConfig) GetProducer() LoggerProducer { return lc.producer }
+func (lc *LoggerConfig) Producer() LoggerProducer { return lc.producer }
+
+// Type returns the type string.
+func (lc *LoggerConfig) Type() string { return lc.info.Type }
+
+// Config returns the raw logger config.
+func (lc *LoggerConfig) Config() RawLoggerConfig { return lc.info.Config }
```

Comment:
is there anyway the interface can return the result of "resolve" rather than a byte slice

Reply:
actually we don't need this method, removing it